fedimint-server/src/consensus/api.rs

```diff
@@ -363,6 +366,75 @@ impl ConsensusApi {
     ) -> Option<ClientBackupSnapshot> {
         dbtx.get_value(&ClientBackupKey(id)).await
     }
+
+    /// List API URL announcements from all peers we have received them from (at
+    /// least ourselves)
+    async fn api_announcements(&self) -> Vec<SignedApiAnnouncement> {
+        self.db
+            .begin_transaction_nc()
+            .await
+            .find_by_prefix(&ApiAnnouncementPrefix)
+            .await
+            .map(|(_, announcement)| announcement)
+            .collect()
+            .await
+    }
+
+    /// Add an API URL announcement from a peer to our database to be returned
+    /// by [`ConsensusApi::api_announcements`].
+    async fn submit_api_announcement(
+        &self,
+        announcement: SignedApiAnnouncement,
+    ) -> Result<(), ApiError> {
+        if self.cfg.local.identity == announcement.api_announcement.peer {
+            return Err(ApiError::bad_request(
+                "Cannot submit own announcement".into(),
+            ));
+        }
+
+        let Some(peer_key) = self
+            .cfg
+            .consensus
+            .broadcast_public_keys
+            .get(&announcement.api_announcement.peer)
+        else {
+            return Err(ApiError::bad_request("Peer not in federation".into()));
+        };
+
+        if !announcement.verify(SECP256K1, peer_key) {
```

Comment:
which is better to verify first, the signature or that the nonce is incremented?

Reply:
Really depends on if the DB or CPU for signature verification is the bottle neck. Both aren't really a huge DoS risk imo.